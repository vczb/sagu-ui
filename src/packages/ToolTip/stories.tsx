import React from 'react'
import { StoryFn, Meta } from '@storybook/react'
import Tooltip, { TooltipProps } from '.'

export default {
  title: 'Tooltip',
  component: Tooltip,
  args: {
    children: 'Hover me',
    content: 'Tooltip text here'
  },
  argTypes: {
    position: {
      control: {
        type: 'select',
        options: ['top', 'bottom', 'left', 'right']
      }
    }
  }
} as Meta

export const Default: StoryFn<TooltipProps> = (args) => <Tooltip {...args} />

// Separated stories for each position
export const Top: StoryFn<TooltipProps> = (args) => (
  <Tooltip {...args} position="top">
    Top
  </Tooltip>
)

export const Bottom: StoryFn<TooltipProps> = (args) => (
  <Tooltip {...args} position="bottom">
    Bottom
  </Tooltip>
)

export const Left: StoryFn<TooltipProps> = (args) => (
  <Tooltip {...args} position="left">
    Left
  </Tooltip>
)

export const Right: StoryFn<TooltipProps> = (args) => (
  <Tooltip {...args} position="right">
    Right
  </Tooltip>
)
