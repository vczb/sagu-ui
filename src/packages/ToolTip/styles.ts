import styled, { css } from 'styled-components'
import { TooltipProps } from '.'

export const Wrapper = styled.div`
  position: relative;
  display: inline-block;
`

export const Content = styled.span<Pick<TooltipProps, 'position'>>`
  ${({ theme, position }) => css`
    visibility: hidden;
    max-width: calc(${theme.container.small} / 2);
    background-color: ${theme.colors.base.black};
    color: ${theme.colors.base.white};
    text-align: center;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xsmall} ${theme.spacings.small};
    position: absolute;
    z-index: ${theme.layers.menu};

    // Dynamic positioning based on the provided position prop
    ${position === 'top' &&
    css`
      bottom: calc(100% + ${theme.spacings.xsmall});
      left: 50%;
      transform: translateX(-50%);
    `}
    ${position === 'bottom' &&
    css`
      top: calc(100% + ${theme.spacings.xsmall});
      left: 50%;
      transform: translateX(-50%);
    `}
    ${position === 'left' &&
    css`
      top: 50%;
      right: calc(100% + ${theme.spacings.small});
      transform: translateY(-50%);
    `}
    ${position === 'right' &&
    css`
      top: 50%;
      left: calc(100% + ${theme.spacings.small});
      transform: translateY(-50%);
    `}

    ${Wrapper}:hover & {
      visibility: visible;
    }
  `}
`
