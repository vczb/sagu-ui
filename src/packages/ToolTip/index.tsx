import React, { ReactNode } from 'react'
import * as S from './styles'

export type TooltipProps = {
  children: ReactNode
  content: ReactNode
  position?: 'top' | 'bottom' | 'left' | 'right'
}

const Tooltip: React.FC<TooltipProps> = ({
  children,
  content,
  position = 'top'
}) => {
  return (
    <S.Wrapper>
      {children}
      <S.Content position={position}>{content}</S.Content>
    </S.Wrapper>
  )
}

export default Tooltip
