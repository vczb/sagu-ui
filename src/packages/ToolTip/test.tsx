/* eslint-disable react/no-children-prop */
import React from 'react'
import { fireEvent, render, screen } from '../../utils/testUtils'

import ToolTip from '.'

describe('<ToolTip/>', () => {
  const tooltipArgs = {
    children: 'Hover me',
    content: 'Tooltip text here'
  }
  it('should render the tooltip text', () => {
    render(<ToolTip {...tooltipArgs} />)

    expect(screen.getByText('Hover me')).toBeInTheDocument()
  })

  it('should render the tooltip content when hovered over the text', async () => {
    render(<ToolTip {...tooltipArgs} />)
    const tooltipText = screen.getByText('Hover me')

    fireEvent.mouseOver(tooltipText)

    await screen.findByText('Tooltip text here')
  })

  it('should render left positioned ToolTip', () => {
    const { container } = render(<ToolTip {...tooltipArgs} position="left" />)
    expect(container).toMatchSnapshot()
  })

  it('should render bottom positioned ToolTip', () => {
    const { container } = render(<ToolTip {...tooltipArgs} position="bottom" />)
    expect(container).toMatchSnapshot()
  })

  it('should render right positioned ToolTip', () => {
    const { container } = render(<ToolTip {...tooltipArgs} position="right" />)
    expect(container).toMatchSnapshot()
  })
})
