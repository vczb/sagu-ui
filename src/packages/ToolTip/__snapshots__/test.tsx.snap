// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToolTip/> should render bottom positioned ToolTip 1`] = `
.c1 {
  position: relative;
  display: inline-block;
}

.c3 {
  visibility: hidden;
  max-width: calc(60rem / 2);
  background-color: #222222;
  color: #F8F7FB;
  text-align: center;
  border-radius: 0.4rem;
  padding: 1.6rem 2.4rem;
  position: absolute;
  z-index: 20;
  top: calc(100% + 1.6rem);
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c0:hover .c2 {
  visibility: visible;
}

<div>
  <div
    class="c0 c1"
  >
    Hover me
    <span
      class="c2 c3"
    >
      Tooltip text here
    </span>
  </div>
</div>
`;

exports[`<ToolTip/> should render left positioned ToolTip 1`] = `
.c1 {
  position: relative;
  display: inline-block;
}

.c3 {
  visibility: hidden;
  max-width: calc(60rem / 2);
  background-color: #222222;
  color: #F8F7FB;
  text-align: center;
  border-radius: 0.4rem;
  padding: 1.6rem 2.4rem;
  position: absolute;
  z-index: 20;
  top: 50%;
  right: calc(100% + 2.4rem);
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c0:hover .c2 {
  visibility: visible;
}

<div>
  <div
    class="c0 c1"
  >
    Hover me
    <span
      class="c2 c3"
    >
      Tooltip text here
    </span>
  </div>
</div>
`;

exports[`<ToolTip/> should render right positioned ToolTip 1`] = `
.c1 {
  position: relative;
  display: inline-block;
}

.c3 {
  visibility: hidden;
  max-width: calc(60rem / 2);
  background-color: #222222;
  color: #F8F7FB;
  text-align: center;
  border-radius: 0.4rem;
  padding: 1.6rem 2.4rem;
  position: absolute;
  z-index: 20;
  top: 50%;
  left: calc(100% + 2.4rem);
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c0:hover .c2 {
  visibility: visible;
}

<div>
  <div
    class="c0 c1"
  >
    Hover me
    <span
      class="c2 c3"
    >
      Tooltip text here
    </span>
  </div>
</div>
`;
